### PR TITLE
Make projection to use event's type to define & run handlers

### PR DESCRIPTION
**Pull request by mpraglowski**

### CODE REVIEW

File: ruby_event_store/lib/ruby_event_store/projection.rb [hunk 1]
```diff
@@ -42,7 +42,7 @@ def current_state
     end
 
     def call(event)
-      handlers.fetch(event.class).(current_state, event)
+      handlers.fetch(event.type).(current_state, event)
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/projection.rb]: 👍

File: ruby_event_store/lib/ruby_event_store/projection.rb [hunk 2]
```diff
@@ -27,7 +27,7 @@ def init(handler)
 
     def when(events, handler)
       Array(events).each do |event|
-        handlers[event] = handler
+        handlers[event.to_s] = handler
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/projection.rb]: why `to_s`?

File: ruby_event_store/lib/ruby_event_store/projection.rb [hunk 3]
```diff
@@ -85,7 +85,7 @@ def start_events(start)
     end
 
     def transition(state, event)
-      handlers[event.class].(state, event)
+      handlers[event.type].(state, event)
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/projection.rb]: why no `fetch` here?

File: ruby_event_store/lib/ruby_event_store/projection.rb (hunk 2 above)
Comment (mpraglowski) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: Because I still want to setup projection with:

```
...
.when(SomeEventClass, -> { ... })
```

Comment (paneq) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: What syntax do you imagine with `Proto`

File: ruby_event_store/lib/ruby_event_store/projection.rb (hunk 3 above)
Comment (mpraglowski) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: Good question - but should it be? or maybe we want to ignore events for which there is no defined handlers?

Currently when you do not specify any stream to process projection works on all events in all streams (`reduce_from_all_streams` method). Current implementation forces you to provide handlers for all domain event types :( Good luck on some production systems with reasonable number of domain events ;)

File: ruby_event_store/lib/ruby_event_store/projection.rb (hunk 2 above)
Comment (mpraglowski) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: AFAIK `Proto` implements `type` - so you could use anything that matches Proto's events types.

Comment (mpraglowski) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: We could not use `event.class` anymore as we want to allow use self defined Event base classes.

File: ruby_event_store/lib/ruby_event_store/projection.rb (hunk 3 above)
Comment (mpraglowski) [reply ruby_event_store/lib/ruby_event_store/projection.rb]:
```
      @handlers = Hash.new { ->(_, _) {} }
```

That's why it's not fetch. I was wrong ... no need to define all kind of events - I've missed that.
`handlers[event.type]` for not existing event type will return "empty" lambda.

File: ruby_event_store/lib/ruby_event_store/projection.rb (hunk 2 above)
Comment (pawelpacana) [reply ruby_event_store/lib/ruby_event_store/projection.rb]: `to_s` actually allows Proto to be used here

**Pull request merged**